python/ray/tests/test_memory_deadlock.py

```diff
@@ -0,0 +1,259 @@
+                first_fraction=memory_usage_threshold_fraction - 0.1,
+                second_fraction=memory_usage_threshold_fraction + 0.25,
+                leaker=leaker,
+                actor_allocation_first=True,
+            )
+        )
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_task_with_nested_actor_with_actor_last(
+    ray_with_memory_monitor,
+):
+    leaker = Leaker.options(max_restarts=0, max_task_retries=0).remote()
+    with pytest.raises(ray.exceptions.OutOfMemoryError) as _:
+        ray.get(
+            task_with_nested_actor.remote(
+                first_fraction=memory_usage_threshold_fraction - 0.1,
+                second_fraction=memory_usage_threshold_fraction + 0.25,
+                leaker=leaker,
+                actor_allocation_first=False,
+            )
+        )
+
+
+# Used for syncing allocations
+@ray.remote
+class BarrierActor:
+    def __init__(self, num_objects):
+        self.barrier = threading.Barrier(num_objects, timeout=30)
+
+    def wait_all_done(self):
+        self.barrier.wait()
+
+
+@ray.remote
+class ActorWithNestedTask:
+    def __init__(self, barrier=None):
+        self.mem = []
+        self.barrier = barrier
+
+    def perform_allocations(self, actor_bytes, nested_task_bytes):
+        self.mem = alloc_mem(actor_bytes)
+        if self.barrier:
+            ray.get(self.barrier.wait_all_done.remote())
+        ray.get(allocate_memory.options(max_retries=0).remote(nested_task_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_actor_with_nested_task(
+    ray_with_memory_monitor,
+):
+    leaker = ActorWithNestedTask.options(max_restarts=0, max_task_retries=0).remote()
+    actor_bytes = get_additional_bytes_to_reach_memory_usage_pct(
+        (memory_usage_threshold_fraction + 0.25) / 2
+    )
+    nested_task_bytes = (
+        get_additional_bytes_to_reach_memory_usage_pct(
+            memory_usage_threshold_fraction + 0.25
+        )
+        - actor_bytes
+    )
+    with pytest.raises(ray.exceptions.RayTaskError) as _:
+        ray.get(leaker.perform_allocations.remote(actor_bytes, nested_task_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_two_sets_of_actor_with_nested_task(
+    ray_with_memory_monitor,
+):
+    barrier = BarrierActor.options(
+        max_restarts=0, max_task_retries=0, max_concurrency=2
+    ).remote(2)
+    leaker1 = ActorWithNestedTask.options(max_restarts=0, max_task_retries=0).remote(
+        barrier
+    )
+    leaker2 = ActorWithNestedTask.options(max_restarts=0, max_task_retries=0).remote(
+        barrier
+    )
+    bytes_first = get_additional_bytes_to_reach_memory_usage_pct(0.25)
```

Comment:
why not do the following

allocate the actors each with (threshold - 0.1) / 2

allocate the nested tasks with mem(threshold) - mem(threshold - 0.2)

Reply:
This particular test behaves strangely. Both of the ray.gets will fail if I do what you suggested, which does not make sense to me

Reply:
Can you check the log to see what is happening? Especially what is printed in worker killing policy?